### PR TITLE
Fix intrinsic

### PR DESCRIPTION
**Pull request by DataCorrupted**

### CODE REVIEW

**Comment by oakrc** [comment]
- Complexity issue is fixed for sure. The compiler chose to use operator bool() for some reason and simdjson didn't like that.
- I haven't finished testing the intrinsic fix yet, but I think it works?
```diff
diff --git a/mapper/src/matchertree.cpp b/mapper/src/matchertree.cpp
index a2d6455..7072f1a 100644
--- a/mapper/src/matchertree.cpp
+++ b/mapper/src/matchertree.cpp
@@ -140,7 +140,8 @@ std::vector<Pattern> getPatterns(ondemand::document &TableJSON) {
     } else {
       ThePattern.PatPredIdx = std::numeric_limits<size_t>::max();
     }
-    ThePattern.Complexity = PatternObject["complexity"];
+    // explicitly cast to prevent using operator bool()
+    ThePattern.Complexity = (int64_t)PatternObject["complexity"];
     ThePattern.IncludePath = PatternObject["path"].get_string().value();
     StringRef SrcDst = PatternObject["pattern"].get_string().value();
     std::tie(ThePattern.Src, ThePattern.Dst) = SrcDst.split(" -> ");
@@ -437,8 +438,19 @@ std::vector<Intrinsic::ID> MatcherTree::blameTargetIntrinsic() const {
       Intrinsic::ID IID = std::stoul(SMatch[1]);
 
       // A few of them are not target intrinsic functions
-      if (Function::isTargetIntrinsic(IID))
+      if (!Function::isTargetIntrinsic(IID))
+        continue;
+
+      if (Intrinsic::isOverloaded(IID)) {
         TargetIIDs.insert(IID);
+        continue;
+      }
+      Intrinsic::ID CurIID = IID + 1;
+      StringRef BaseName = Intrinsic::getBaseName(IID);
+      while (Intrinsic::getBaseName(CurIID).starts_with(BaseName) &&
+             !Intrinsic::isOverloaded(CurIID)) {
+        TargetIIDs.insert(CurIID++);
+      }
     }
   }
   return std::vector<Intrinsic::ID>(TargetIIDs.begin(), TargetIIDs.end());
```

**Comment by oakrc** [comment]
Sure, I'm moving the second patch to InsertIntrinsicStrategy.

File: mutator/src/InsertIntrinsicStrategy.cpp [hunk 1]
```diff
@@ -21,7 +21,15 @@ Function *InsertIntrinsicStrategy::chooseFunction(Module *M,
                                                   RandomIRBuilder &IB) {
   LazyInit();
   auto RS = makeSampler(IB.Rand, IIDs);
-  return Intrinsic::getDeclaration(M, RS.getSelection());
+  if (RS.isEmpty())
+    return nullptr;
+  Intrinsic::ID IID = RS.getSelection();
+  if (!Intrinsic::isOverloaded(IID)) {
+    return Intrinsic::getDeclaration(M, IID);
```
Comment (DataCorrupted) [inline mutator/src/InsertIntrinsicStrategy.cpp:28]: I think InsertIntrinsicStrategy has a pretty decent fix (temporary) fixing for now.

**Comment by DataCorrupted** [comment]
Also, you can directly push to this branch if you want to.

File: mutator/src/InsertIntrinsicStrategy.cpp (hunk 1 above)
Comment (oakrc) [reply mutator/src/InsertIntrinsicStrategy.cpp:28]: Sure, but later if you want I can handle non-overloaded functions in a similar way as the second patch.

**Review by oakrc — APPROVED** [review]

**Pull request merged**